polar/src/polar.lalrpop

```diff
@@ -120,79 +130,42 @@ Object<T>: BTreeMap<Symbol, Term> = {
-
-InstanceLiteral<T>: Term = <start:@L> <tag:Name> <fields:Object<T>> => {
-    let instance = InstanceLiteral{tag, fields: Dictionary{fields}};
-    Term::new_from_parser(src_id, start, Value::InstanceLiteral(instance))
+DictionaryPattern: Value = <fields:Object<Exp9<"Pattern">>> => {
+    let dict = Dictionary{fields};
+    Value::Pattern(Pattern::Dictionary(dict))
 };
 
-pub InstanceLiteralTerm: Term = <InstanceLiteral<Exp5<"Term">>> => <>;
-pub InstanceLiteralPattern: Term = <InstanceLiteral<Exp9<"Pattern">>> => <>;
-
-// Either a *rest variable, or an ordinary term.
-RestVar<T>: Term = {
-    <start:@L> "*" <rest:Name> => {
-        Term::new_from_parser(src_id, start, Value::RestVariable(rest))
-    },
-    <Term> if T == "Term" => <>,
-    <Pattern> if T == "Pattern" => <>
+InstanceLiteralTerm: Value = <tag:Name> <fields:Object<Exp5<"Term">>> => {
+    let instance = InstanceLiteral{tag, fields: Dictionary{fields}};
+    Value::InstanceLiteral(instance)
 };
 
-pub Pattern: Term = {
-    <term:InstanceLiteralPattern> => Pattern::term_as_pattern(&term),
-    <term:DictionaryPattern> => Pattern::term_as_pattern(&term),
-    <term:Number> => term,
-    <term:PolarString> => term,
-    <term:Boolean> => term,
-    <term:Symbol> => term,
-    <start:@L> "[" <terms:(<Pattern> ",")*> <term:RestVar<"Pattern">?> "]" => {
-        match term {
-            Some(term) => Term::new_from_parser(src_id, start, Value::List(vec![term])),
-            None => Term::new_from_parser(src_id, start, Value::List(terms))
-        }
-    }
+InstanceLiteralPattern: Value = <tag:Name> <fields:Object<Exp9<"Pattern">>> => {
+    let instance = InstanceLiteral{tag, fields: Dictionary{fields}};
+    Value::Pattern(Pattern::Instance(instance))
 };
 
-pub Term: Term = {
-    <term:Number> => term,
-    <term:PolarString> => term,
-    <term:Boolean> => term,
-    <term:Symbol> => term,
-    <term:DictionaryTerm> => term,
-    <start:@L> <op:BuiltinOperation> => Term::new_from_parser(src_id, start, Value::Expression(op)),
-    <start:@L> <op:RewrittenOperation> => Term::new_from_parser(src_id, start, Value::Expression(op)),
-    <start:@L> <pred:Predicate> => Term::new_from_parser(src_id, start, Value::Call(pred)),
-    <start:@L> "[" <mut terms:(<Term> ",")*> <term:RestVar<"Term">?> "]" => {
-        match term {
-            Some(term) => {
-                terms.push(term);
-                Term::new_from_parser(src_id, start, Value::List(terms))
-            },
-            None => Term::new_from_parser(src_id, start, Value::List(terms))
-        }
-    }
-};
+// ****** Operations ******* //
 
 BuiltinOperator: Operator = {
     "debug" => Operator::Debug,
     "print" => Operator::Print,
 };
 
-pub BuiltinOperation: Operation = {
-    <op:BuiltinOperator> "(" <mut args:(<Exp2<"Term">> ",")*> <arg:Exp2<"Term">?> ")" => {
+BuiltinOperation: Value = {
+    <op:BuiltinOperator> "(" <mut args:(<TermExp> ",")*> <arg:TermExp?> ")" => {
```

Comment:
I'm not sure this should be a full `TermExp` as the argument.  Is `print(x in 1)` valid? Or `print(1 = x)`? Or even `print(y = 1 and x = 1)`

Those would all parse with this change.

Reply:
They already do:

```
query> print(x in 1)
x in 1
True
```

Exp2 is already matching everything but `x and y`. It's not really achieving much though, you're right